tests/test_checkout.py

```diff
@@ -8,64 +8,207 @@ class TestCheckout(object):
         co.close()
 
     def test_write_after_repo_deletion(self, written_repo, array5by7):
         co = written_repo.checkout(write=True)
         shutil.rmtree(written_repo._repo_path)
         with pytest.raises(OSError):
             co.datasets.init_dataset('dset', prototype=array5by7)
+        co.metadata.add('a', 'b')
         dset = co.datasets['_dset']
         dset['1'] = array5by7
-        # todo: shouldn't it be atleast same as above
         with pytest.raises(FileNotFoundError):
             co.commit()
         co.close()
 
-
-@pytest.mark.skip(reason='not implemented')
-class TestBranching(object):
-
-    def test_merge(self):
-        pass
-
-    def test_name_conflict(self):
-        pass
-
-    def test_delete_branch(self):
-        pass
-
-    def test_merge_multiple_checkouts(self):
-        pass
-
-    def test_merge_diverged(self):
-        pass
-
-    def test_merge_diverged_conflict(self):
-        pass
+    def test_operate_on_dataset_after_closing_old_co(self, repo, array5by7):
+        co = repo.checkout(write=True)
+        dset = co.datasets.init_dataset('dset', prototype=array5by7)
+        co.commit()
+        co.close()
+        co = repo.checkout(write=True)
+        # TODO: what is the design here. IMO, the old instance should not execute
+        dset.add(array5by7, '1')
+        co.commit()
+        co.close()
+        # TODO: even after the close, I could still fetch the
+        #   data from the old dset object.
+        print(dset['1'])
+
+    def test_operate_on_closed_checkout(self, repo, array5by7):
+        co = repo.checkout(write=True)
+        co.datasets.init_dataset('dset', prototype=array5by7)
+        co.commit()
+        co.close()
+        # TODO: This should raise permission error or something
+        with pytest.raises(KeyError):
+            co.datasets['dset']['1'] = array5by7
+        co.metadata.add('a', 'b')
+
+
+class TestBranching:
+
+    def test_invalid_name(self, repo):
+        with pytest.raises(ValueError):
+            repo.create_branch('dummy branch')
+
+    def test_new_branch_from_new_repo(self, repo):
+        branch = repo.create_branch('testbranch')
+        # TODO: it doesnt raise value error to the user
+        # instead return false and print out the value error
+        # Ideally, it should let users create new branch
+        co = repo.checkout(write=True, branch_name=branch)
+
+    def test_merge(self, written_repo, array5by7):
+        branch = written_repo.create_branch('testbranch')
+        assert type(branch) is str
+        co = written_repo.checkout(write=True, branch_name=branch)
+        assert co._branch_name == branch
+        co.datasets['_dset']['1'] = array5by7
+        co.metadata.add('a', 'b')
+        co.commit()
+        co.close()
+        written_repo.merge('test merge', 'master', branch)
+        co = written_repo.checkout()
+        # TODO: It's good to give access to name of the branch from Readcheckout
```

Comment:
I agree in principle, but disagree in practice. Since a read checkout and write checkout can operate at the same time on the same repository, we should only provide the commit which the reader checkout is operating on, not the actual branch name. 

Imaging the following scenario where a reader checks out `branch "A"` which (at the time of checkout) has head commit `2`. 

```
                         Reader c/o 
0 ---------> 1
              \___________> 2
                         branch A
```

Then a writer check's out `branch "A"` as well

```
                           Reader c/o
0 ---------> 1             Writer c/o
              \___________> 2
                         branch A
```

And then (while the reader is still active) the writer makes `commit 3` on `branch "A"`

```
                        Reader c/o
0 ---------> 1                                next writer c/o
            \___________-> 2 _________________-> 3
                        branch A
```

If we provided the "branch name" to the reader checkout object, it would be defined when it was first created, but then could potentially change to a `null` value over the course of it's lifetime... this would be a crazy behavior for user defined code which might depend on a branch name.

If we only provide the commit hash, we can be sure that it will never change under any circumstances, and a user can actually write application code which depends on that value for whatever need they might have.

Reply:
This makes a lot of sense. Although, I have a slightly convoluted solution where we raise a warning if the head moved further when the user tries to fetch the branch name from a read checkout and return the branch name. If the head not moved, return the branch name without raising the warning. I am just thinking aloud and perhaps this could be a problem for another day.

Reply:
So I double checked, and this information is already provided to each of the checkout objects in one of their underscore attributes:

https://github.com/tensorwerk/hangar-py/blob/065ae0735c630f627fb5f5d9cd634739408a5733/src/hangar/checkout.py#L42

As part of #32 I'll make a `commit` attribute available to the user (following the descriptor protocol to prevent a user setting this value to something else after construction)

I'll do the same for the `branch_name` attribute for a write-enabled checkout.

https://github.com/tensorwerk/hangar-py/blob/065ae0735c630f627fb5f5d9cd634739408a5733/src/hangar/checkout.py#L124